server/container_create.go

```diff
@@ -1056,10 +1066,126 @@ func (s *Server) createSandboxContainer(ctx context.Context, containerID string,
 		}
 	}
 
-	// Force host mapping for the container's content.
-	idmapOptions := &storage.IDMapOptions{
-		HostUIDMapping: true,
-		HostGIDMapping: true,
+	// Decide which user namespace to use and what ID mappings we should specify for the container's storage.
+	var idmapOptions *storage.IDMapOptions
+	userNamespaceMode := pb.NamespaceMode_NODE
+	if linux != nil {
+		// If we ever get User namespace options, this should probably key off of that.  In the
+		// meantime, if we're privileged, use the host user namespace.  Otherwise, use the pod's
+		// user namespace, unless we were passed annotations telling us which mappings to use,
+		// in which case we'll set up a user namespace for just the container.
+		if !privileged {
+			userNamespaceMode = pb.NamespaceMode_POD
+			if kubeAnnotations[annotations.UIDMappings] != "" || kubeAnnotations[annotations.GIDMappings] != "" {
+				userNamespaceMode = pb.NamespaceMode_CONTAINER
+			}
+		}
+		switch userNamespaceMode {
+		case pb.NamespaceMode_NODE:
+			// We're going to use the host user namespace, so use no ID mappings.
+			if err := specgen.RemoveLinuxNamespace(string(rspec.UserNamespace)); err != nil {
+				return nil, err
+			}
+			idmapOptions = &storage.IDMapOptions{
+				HostUIDMapping: true,
+				HostGIDMapping: true,
+			}
+		case pb.NamespaceMode_POD:
+			// Use the sandbox's user namespace and ID mappings.
+			userns, err := sb.UserNsPath()
+			if err != nil {
+				if !os.IsNotExist(err) {
+					return nil, fmt.Errorf("error determining path of user namespace for pod: %v", err)
+				}
+				empty := ""
+				userns = &empty
+			}
+			if userns == nil || *userns == "" {
+				// The sandbox is also in the host user namespace.
+				if err := specgen.RemoveLinuxNamespace(string(rspec.UserNamespace)); err != nil {
+					return nil, err
+				}
+				idmapOptions = &storage.IDMapOptions{
+					HostUIDMapping: true,
+					HostGIDMapping: true,
+				}
+			} else {
+				// The sandbox is in its own user namespace -- configure to use its ID mappings.
+				if err := specgen.AddOrReplaceLinuxNamespace(string(rspec.UserNamespace), *userns); err != nil {
+					return nil, err
+				}
+				idmapOptions = &storage.IDMapOptions{
+					UIDMap: sb.UIDMap(),
+					GIDMap: sb.GIDMap(),
+				}
+			}
+		case pb.NamespaceMode_CONTAINER:
```

Comment:
Is this something that is even supported?  Separate UserNS for containers in the same pod?